sdk/dotnet/Pulumi.Tests/Serialization/ResourceRefPropertyTests.cs

```diff
@@ -0,0 +1,324 @@
+            Assert.NotNull(stack);
+
+            var values = (await stack.Values.DataTask).Value;
+            Assert.Equal(values["expectedUrn"], values["actualUrn"]);
+        }
+
+        [Theory]
+        [InlineData(false)]
+        [InlineData(true)]
+        public async Task SerializeComponentResource(bool isPreview)
+        {
+            var resources = await RunAsync<MyStack>(isPreview);
+            var res = resources.OfType<MyComponentResource>().FirstOrDefault();
+            Assert.NotNull(res);
+
+            var urn = (await res.Urn.DataTask).Value;
+
+            var v = await SerializeToValueAsync(res);
+            Assert.Equal(CreateComponentResourceReference(urn), v);
+        }
+
+        [Theory]
+        [InlineData(false)]
+        [InlineData(true)]
+        public async Task SerializeComponentResourceDownlevel(bool isPreview)
+        {
+            var resources = await RunAsync<MyStack>(isPreview);
+            var res = resources.OfType<MyComponentResource>().FirstOrDefault();
+            Assert.NotNull(res);
+
+            var urn = await SerializeToValueAsync(res.Urn);
+
+            var v = await SerializeToValueAsync(res, false);
+            Assert.Equal(urn, v);
+        }
+
+        public class DeserializeComponentResourceStack : Stack
+        {
+            [Output("values")]
+            public Output<ImmutableDictionary<string, string>> Values { get; private set; } = null!;
+
+            public DeserializeComponentResourceStack()
+            {
+                var res = new MyComponentResource("test", null, null);
+
+                var urn = res.Urn.DataTask.Result.Value;
+
+                var v = DeserializeValue<MyComponentResource>(CreateComponentResourceReference(urn));
+
+                this.Values = Output.Create(new Dictionary<string, string> {
+                    { "expectedUrn", urn },
+                    { "actualUrn", v.Urn.DataTask.Result.Value },
+                }.ToImmutableDictionary());
+            }
+        }
+
+        [Theory]
+        [InlineData(false)]
+        [InlineData(true)]
+        public async Task DeserializeComponentResource(bool isPreview)
+        {
+            var resources = await RunAsync<DeserializeComponentResourceStack>(isPreview);
+
+            var stack = resources.OfType<DeserializeComponentResourceStack>().FirstOrDefault();
+            Assert.NotNull(stack);
+
+            var values = (await stack.Values.DataTask).Value;
+            Assert.Equal(values["expectedUrn"], values["actualUrn"]);
+        }
+
+        public class DeserializeMissingComponentResourceStack : Stack
+        {
+            [Output("values")]
+            public Output<ImmutableDictionary<string, string>> Values { get; private set; } = null!;
+
+            public DeserializeMissingComponentResourceStack()
+            {
+                var res = new MissingComponentResource("test", null, null);
+
+                var urn = res.Urn.DataTask.Result.Value;
```

Comment:
I guess it's a nit for a test but you could assign `Values` with `Output.Tuple` machinery and avoid using `.Result`'s.